### PR TITLE
UI Events. Buttons, hovering, clicks clicks CLICKS!

### PR DESCRIPTION
**Pull request by jojolepro**
Much hovering, much clicking!

Review-ready, except if @Xaeroxe wants more features in this PR ;)

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/580)

### CODE REVIEW

File: amethyst_ui/src/lib.rs [hunk 1]
```diff
@@ -1,10 +1,11 @@
 //! Provides components and systems to create an in game user interface.
 
-#![deny(missing_docs)]
+//#![deny(missing_docs)]
```
Comment (Xaeroxe) [inline amethyst_ui/src/lib.rs]: Please uncomment this.

Comment (torkleyy) [reply amethyst_ui/src/lib.rs]: Or even better, change it to `warn`.

File: amethyst_ui/src/event.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        pos_x > rect_x && pos_x < rect_x + width && pos_y > rect_y && pos_y < rect_y + height
+    }
+}
+
+impl<'a> System<'a> for UiMouseSystem {
+    type SystemData = (
+        Entities<'a>,
+        ReadStorage<'a, UiTransform>,
+        ReadStorage<'a, MouseReactive>,
+        Fetch<'a, InputHandler<String, String>>,
```
Comment (Xaeroxe) [inline amethyst_ui/src/event.rs]: `String` should be generic here.

File: amethyst_ui/src/lib.rs (hunk 1 above)
Comment (jojolepro) [reply amethyst_ui/src/lib.rs]: done

File: amethyst_ui/src/event.rs (hunk 2 above)
Comment (jojolepro) [reply amethyst_ui/src/event.rs]: done

**Comment by Rhuagh** [comment]
:lgtm:

I'd probably do things a bit differently, but this should work fine!

---

Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/amethyst/amethyst/580#-:-L5zN_O04BKbKf6Z6KZd:bxgnj4i)*

**Comment by torkleyy** [comment]
Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/580)*

**Comment by Rhuagh** [comment]
You'll have to fix so it builds before we merge :P

**Comment by jojolepro** [comment]
haha I just saw. Will do ;)

**Comment by Xaeroxe** [comment]
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/580)*

**Comment by Rhuagh** [comment]
Squash pls.

**Pull request merged**